---
title: "Day 2 - Module 08: Count normalizations and DESeq2 model fitting"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: true
            fig_caption: true
            markdown: GFM
            code_download: true
---
<style type="text/css">
body{ /* Normal  */
      font-size: 14pt;
  }
pre {
  font-size: 12pt
}
code.r{
  font-size: 12pt;
}
</style>

<!--- Allow the page to be wider --->
<style>
    body .main-container {
        max-width: 1200px;
    }
</style>

> # Objectives:    
> * Discuss count normalizations
> * Execute model fitting for differential expression comparisons

```{r LoadRunningData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
load("rdata/RunningData.RData")
```


# Differential Expression Workflow

Here we will proceed with count normalizations and fit our DESeq2 model.

![](./images/wayfinder/wayfinder-04.png){width=75%}

---

# Count normalizations

Since counts of mapped reads for each gene is proportional to the expression of RNA in addition to many “uninteresting” other factors, normalization is the process of scaling raw count values to account for the “uninteresting” factors and ensure expression levels are more comparable.

## Normalization goals

Two common factors that need to be accounted for during normalization are **sequencing depth** and **gene length**. Common normalization approaches (such as FPKM, RPKM, CPM, TPM, etc.) account for one or both of these factors.    

* **Sequencing depth** normalization is necessary to account for the proportion of reads per gene expected for more deeply sequenced samples (like in pink below) versus a less deeply sequenced sample (like in green below).    

![*Note that each pink or green rectangle represents an aligned read, with reads spanning an intron connected by a dashed line. Figure from [HBC  training materials](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html)*](./images/normalization_methods_depth.png)    


* **Gene length** normalization is necessary since genes of different lengths have different probabilities of generating fragments that end up in the library. In the example below, both genes have similar levels of expression. However, the longer gene (Gene X) will have more mapped reads than the short gene (Gene Y).    

![*Each pink rectangle represents an aligned read, with reads spanning an intron connected by a dashed line. Figure from [HBC  training materials](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html)*](./images/normalization_methods_length.png){width=75%}      


> **Note**: The above figures are originally from a [HBC  tutorial](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html) that also includes a detailed comparison of different normalization (CPM, TPM, FPKM) approaches and their best uses.

**Check-in: Questions about normalizations?**

## DESeq2 normalizations

DESeq2 has an [internal normalization process](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106) that accounts for **RNA composition**. A few highly differentially expressed genes, differences in the number of genes expressed between samples, or contamination are not accounted for by depth or gene length normalization methods. Accounting for RNA composition is particularly important for differential expression analyses, regardless of the tool used.

For data exploration and visualizations, it is helpful to generate an object of independently normalized counts. We will use the [rlog transformation](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations) that accounts for sequencing depth for each sample and RNA composition differences in the overall experiment for downstream quality control visualizations.

The rlog transformation produces log2 scale data that has also been normalized to overall library size as well as variance across genes at different mean expression levels. For larger numbers of samples, there is an alternative transformation method, [vst](http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations) that can be used instead for count normalizations.


The command to generate the normalized count object has a few parts, including `dds` as an input and providing a value to the option `blind`.
```{r CountNorm}
rld <- rlog(dds, blind = TRUE)
```

*Note: We see a message here that our data did not fit the default 'parametric' dispersion model so a local regression was substituted. If we had more time, we might look at a dispersion plot with the `plotDispEsts(dds)` function, but as [this bioconductor thread](https://support.bioconductor.org/p/107937/) discusses, other visualizations of our data might be more helpful and/or easier to interpret.

Next, we'll look at the results of the transformation.
```{r CountNormCheck}
head(assay(rld), 2)
```

Looking at the rld values, we can see that they are now in log scale. Since we set `blind=TRUE`, the transformation is blind to the sample information we specified in the design formula. The normalized counts are helpful for visualization methods during expression-level quality assessment but **aren't used in the model fitting**.

# DESeq2 Model Fitting

Next, we'll fit our standard model using the `DESeq` function and take a look at the objects we generate.
```{r FitModelStandard, message=FALSE}
dds <- DESeq(dds)
```

This command applies the model to our data, using the sample information supplied when generating the `dds` object so it takes some time to run.

```{r FitModelStandardCheck1, eval=TRUE}
resultsNames(dds)
```

The results include three pairwise comparisons to the specified control as default but other information is now stored in the `dds` object so we can generate additional pairwise comparisons.

```{r FitModelStandardCheck2, eval=FALSE}
head(dds)
```

**Checkpoint**: *If you see the same results when you execute `resultsNames(dds)`, please indicate with the green 'yes' button. Otherwise, please use the red 'x' button to get help before the break*



#### [Optional Exercise]

If you executed the commands in the option exercise from Module 07, you can fit a seperate DESeq2 model for the patient sample example, which included a covariate in our model.

<details>
    <summary>*Click for fitting a model that includes a covariate*</summary>
    ```{r FitModelCovariate, eval= FALSE, message=FALSE}
    dds_patient <- DESeq(dds_patient)
    ```
    ```{r FitModelCovariateCheck1,eval=FALSE}
    resultsNames(dds_patient)
    ```
    If you run through the optional exercises, you can explore the impact of adding a covariate by substituting `dds_patient` for `dds` and re-running those commands since both DESeq2 objects have their data organized in the same way.
</details>




# Summary

In this section, we:   

* Learned about count normalizations and uses    
* Generated a normalized count table    
* Fit two DESeq2 models for our data    
* Saw the impact of including a covariate in our model



---

# Sources
## Training resources used to develop materials
* HBC DGE setup: https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html
* HBC Count Normalization: https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html
* DESeq2 standard vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
* DESeq2 beginners vignette: https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf
* Bioconductor RNA-seq Workflows: https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html



```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
#Hidden code block to write out data for knitting
save.image(file = "rdata/RunningData.RData")

```


These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.
