<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />


<title>Day 1 - Module 03: Reference Genomes</title>

<script src="site_libs/header-attrs-2.10/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02_QC.html">Quality Control</a>
    </li>
    <li>
      <a href="Module02optional_Cutadapt_MultiQC.html">Optional: Cutadapt/MultiQC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module04_Alignment.html">Alignment/Quantification (STAR/RSEM)</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module05_MultiQC_and_Count_Matrix.html">Count matrix / Alignment QC</a>
    </li>
    <li>
      <a href="Module05optional_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">Introduction to DESeq2</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DESeq2 Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">Count Normalizations and DESeq2 model fitting</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample Visualizations for QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Comparisons and Gene Annotations</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Day 1 - Module 03: Reference Genomes</h1>
<h4 class="author">UM Bioinformatics Core</h4>

</div>


<style type="text/css">
body{ /* Normal  */
      font-size: 14pt;
  }
pre {
  font-size: 12pt
}
</style>
<div id="reference-genomes" class="section level1" number="1">
<h1 number="1"><span class="header-section-number">1</span> Reference Genomes</h1>
<p>In this module, we will learn:</p>
<ul>
<li>what a reference genome is and what it contains</li>
<li>details about the FASTA and GTF formats</li>
<li>to appreciate the differences in gene identifiers</li>
<li>how to download a reference genome</li>
</ul>
</div>
<div id="differential-expression-workflow" class="section level1" number="2">
<h1 number="2"><span class="header-section-number">2</span> Differential Expression Workflow</h1>
<p>Here we will set the stage for the next steps by discussing reference genomes, which are integral to genome alignments and gene/isoform quantification. Along the way we will touch on some things to be aware of.</p>
<p><img src="images/wayfinder/wayfinder-02.png" /> <br> <br> <br> <br></p>
</div>
<div id="reference-genomes-1" class="section level1" number="3">
<h1 number="3"><span class="header-section-number">3</span> Reference Genomes</h1>
<p>A reference genome consists of the <strong>reference sequence</strong> and, optionally, any number of <strong>genomic annotations</strong> that describe attributes about that sequence. Examples of annotations include:</p>
<ul>
<li>Gene models consisting of the location and other information about genes.</li>
<li>Variants consisting of the location of common or rare genetic variants, their alleles, and frequencies.</li>
<li>Small RNAs consisting of the location and other information about various types of small RNAs.</li>
</ul>
<p>Of particular relevance to us for this workshop are the reference sequence and gene models.</p>
<div id="reference-sequence" class="section level2" number="3.1">
<h2 number="3.1"><span class="header-section-number">3.1</span> Reference Sequence</h2>
<p>Reference sequence is stored in <a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a> files. They are similar to FASTQ files in their storage of sequence information, but their format is a little different in a couple ways:</p>
<ol style="list-style-type: decimal">
<li>Records are separated by lines beginning with <code>&gt;</code> instead of <code>@</code>.</li>
<li>Only the sequence is stored in a FASTA file, there is no notion of quality attached to the nucleotides.</li>
</ol>
<pre><code>&gt;chrM
GATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCAT
TTGGTATTTTCGTCTGGGGGGTGTGCACGCGATAGCATTGCGAGACGCTG
GAGCCGGAGCACCCTATGTCGCAGTATCTGTCTTTGATTCCTGCCTCATT
CTATTATTTATCGCACCTACGTTCAATATTACAGGCGAACATACCTACTA
AAGTGTGTTAATTAATTAATGCTTGTAGGACATAATAATAACAATTGAAT
GTCTGCACAGCCGCTTTCCACACAGACATCATAACAAAAAATTTCCACCA
AACCCCCCCCTCCCCCCGCTTCTGGCCACAGCACTTAAACACATCTCTGC
CAAACCCCAAAAACAAAGAACCCTAACACCAGCCTAACCAGATTTCAAAT
TTTATCTTTAGGCGGTATGCACTTTTAACAGTCACCCCCCAACTAACACA</code></pre>
</div>
<div id="gene-models" class="section level2" number="3.2">
<h2 number="3.2"><span class="header-section-number">3.2</span> Gene Models</h2>
<p>Well-characterized organisms (e.g. human, mouse, zebrafish) have fairly mature gene models. These are stored in <a href="https://uswest.ensembl.org/info/website/upload/gff.html">GTF</a> format, which gives location and other information about each gene feature. Below are two examples:</p>
<pre><code>chr1    unknown exon    11874   12227   .   +   .   gene_id &quot;DDX11L1&quot;; gene_name &quot;DDX11L1&quot;; transcript_id &quot;NR_046018&quot;; tss_id &quot;TSS16932&quot;;
chr1    unknown exon    12613   12721   .   +   .   gene_id &quot;DDX11L1&quot;; gene_name &quot;DDX11L1&quot;; transcript_id &quot;NR_046018&quot;; tss_id &quot;TSS16932&quot;;
chr1    unknown exon    13221   14409   .   +   .   gene_id &quot;DDX11L1&quot;; gene_name &quot;DDX11L1&quot;; transcript_id &quot;NR_046018&quot;; tss_id &quot;TSS16932&quot;;
chr1    unknown exon    14362   14829   .   -   .   gene_id &quot;WASH7P&quot;; gene_name &quot;WASH7P&quot;; transcript_id &quot;NR_024540&quot;; tss_id &quot;TSS8568&quot;;


1   havana  gene    11869   14409   .   +   .   gene_id &quot;ENSG00000223972&quot;; gene_version &quot;5&quot;; gene_name &quot;DDX11L1&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;transcribed_unprocessed_pseudogene&quot;;
1   havana  transcript  11869   14409   .   +   .   gene_id &quot;ENSG00000223972&quot;; gene_version &quot;5&quot;; transcript_id &quot;ENST00000456328&quot;; transcript_version &quot;2&quot;; gene_name &quot;DDX11L1&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;transcribed_unprocessed_pseudogene&quot;; transcript_name &quot;DDX11L1-202&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;lncRNA&quot;; tag &quot;basic&quot;; transcript_support_level &quot;1&quot;;
1   havana  exon    11869   12227   .   +   .   gene_id &quot;ENSG00000223972&quot;; gene_version &quot;5&quot;; transcript_id &quot;ENST00000456328&quot;; transcript_version &quot;2&quot;; exon_number &quot;1&quot;; gene_name &quot;DDX11L1&quot;; gene_source &quot;havana&quot;; gene_biotype &quot;transcribed_unprocessed_pseudogene&quot;; transcript_name &quot;DDX11L1-202&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;lncRNA&quot;; exon_id &quot;ENSE00002234944&quot;; exon_version &quot;1&quot;; tag &quot;basic&quot;; transcript_support_level &quot;1&quot;;</code></pre>
<p>The GTF format stores specific information in each column:</p>
<table>
<colgroup>
<col width="35%" />
<col width="64%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">1</td>
<td>Chromosome</td>
</tr>
<tr class="even">
<td align="center">2</td>
<td>Source, e.g. ensembl, havana</td>
</tr>
<tr class="odd">
<td align="center">3</td>
<td>Gene feature, e.g. exon, intron, mRNA, transcript</td>
</tr>
<tr class="even">
<td align="center">4</td>
<td>Start location, 1-based</td>
</tr>
<tr class="odd">
<td align="center">5</td>
<td>End location, 1-based</td>
</tr>
<tr class="even">
<td align="center">6</td>
<td>Score</td>
</tr>
<tr class="odd">
<td align="center">7</td>
<td>Strand</td>
</tr>
<tr class="even">
<td align="center">8</td>
<td>Frame, relating to codons</td>
</tr>
<tr class="odd">
<td align="center">9</td>
<td>Attribute, a semicolon separated list of key/value pairs giving additional information about the feature.</td>
</tr>
</tbody>
</table>
</div>
<div id="minutiae-very-briefly" class="section level2" number="3.3">
<h2 number="3.3"><span class="header-section-number">3.3</span> Minutiae, Very Briefly</h2>
<p>Bioinformatics is a relatively new, fast-changing, field and its data standards and formats are no different. Consequently there are some oddities and tedious items of note which we would like to only briefly touch on here.</p>
<div id="genome-builds" class="section level3" number="3.3.1">
<h3 number="3.3.1"><span class="header-section-number">3.3.1</span> Genome Builds</h3>
<p>On occassion new reference genomes are released, and the genome build number changes. You may be familiar with the UCSC manner of naming human genome builds: hg18, hg19, hg38. ENSEMBL, naturally, has their own way of referring to genome builds: GRCh36, GRCh37, and GRCh38. Notice with the most recent human reference, the numbering now aligns between UCSC and ENSEMBL.</p>
<p>Different organisms have their own versioning.</p>
</div>
<div id="annotation-sources" class="section level3" number="3.3.2">
<h3 number="3.3.2"><span class="header-section-number">3.3.2</span> Annotation Sources</h3>
<p><a href="https://www.ncbi.nlm.nih.gov/refseq/rsg/">NCBI RefSeq</a>, <a href="https://www.ensembl.org/info/genome/genebuild/index.html">ENSEMBL</a>, and <a href="https://academic.oup.com/bioinformatics/article/22/9/1036/200093">UCSC Known Genes</a> are the three primary gene annotation databases (different organisms have their own databases). We will not go into exactly how the gene annotations are different, but we note that the are, and <a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-015-1308-8">others have examined the consequences of this</a>.</p>
</div>
<div id="gene-ids" class="section level3" number="3.3.3">
<h3 number="3.3.3"><span class="header-section-number">3.3.3</span> Gene IDs</h3>
<p>The two GTF examples above highlight different ways of referring to the same gene. In the first GTF we see:</p>
<ul>
<li>DDX11L1, the gene symbol, controlled by the <a href="https://www.genenames.org/">Human Gene Nomenclature Committee (HUGO)</a>.</li>
<li>NR_046018, the RefSeq transcript ID</li>
</ul>
<p>And in the second GTF we see:</p>
<ul>
<li>ENSG00000223972, the ENSEMBL gene ID</li>
<li>DDX11L1, the gene symbol, thankfully the same</li>
<li>ENST00000456328, the ENSEMBL transcript ID</li>
</ul>
<p>Translating between different gene IDs is possible, as we will see in Day Two with <code>biomaRt</code>. But in terms of <strong>best practice</strong> it is generally a good idea to avoid using the gene symbol as the primary gene identifier because not everyone refers to the same gene by the same symbol.</p>
</div>
</div>
</div>
<div id="getting-a-reference-genome" class="section level1" number="4">
<h1 number="4"><span class="header-section-number">4</span> Getting a Reference Genome</h1>
<p>The <a href="https://support.illumina.com/sequencing/sequencing_software/igenome.html">Illumina iGenomes</a> resource is one of the easiest, and most comprehensive, ways to download a reference genome. iGenomes includes both the reference sequence and gene models.</p>
<p>Reference genomes can be <strong>very large</strong>, depending on the organism, and so we will not download one to the Amazon instance we are using for this workshop. We’ve included instructions for downloading these, in case you want to download these to the server where you intend to later do a similar RNA-seq analysis (e.g. on High-Performance Compute, GreatLakes).</p>
<details>
<summary>
How would I download references with iGenomes?
</summary>
<p>As noted, it’s not recommended to download the iGenomes references to the AWS instance. However, if you wanted to know in general how you would do that, the process is described here.</p>
<p>First go to the <a href="https://support.illumina.com/sequencing/sequencing_software/igenome.html">iGenomes</a> page, find the build you want from the source you want, right click the genome build you want to download, and select “Copy link location”:</p>
<div class="figure">
<img src="images/genome_copy_link.png" alt="" />
<p class="caption">iGenomes image for copying link location</p>
</div>
<p>Then on the remote server you would go to the directory you’d like to download the genome to and type (that URL is what we copied):</p>
<pre><code>$ wget http://igenomes.illumina.com.s3-website-us-east-1.amazonaws.com/Homo_sapiens/NCBI/GRCh38/Homo_sapiens_NCBI_GRCh38.tar.gz</code></pre>
<p>After the download finishes (it may take a while as it is tens of GB large), you can unpack it with:</p>
<pre><code>$ tar -xf Homo_sapiens_NCBI_GRCh38.tar.gz</code></pre>
</details>
<div id="which-reference-is-right-for-me" class="section level2" number="4.1">
<h2 number="4.1"><span class="header-section-number">4.1</span> Which Reference is Right for Me?</h2>
<p>The key is to be consistent in your research. Switching from ENSEMBL to UCSC will create many headaches because of the change in gene identifiers, and differences in the gene models themselves. Often people choose the one they’re most comfortable with, which is often a function of historical accident. The key is not to overthink it.</p>
<p>Another important note is not to mix the sources. If you download reference sequence from UCSC, don’t use an ENSEMBL GTF (and vice versa). One of the quirky differences between the two databases is that ENSEMBL refers to chromosome only by their number, i.e. <code>1</code>, whereas UCSC refers to chromsomes as <code>chr1</code>. This makes reference FASTAs from one source incompatible with gene builds from another.</p>
<hr />
<p>These materials have been adapted and extended from materials created by the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. These are open access materials distributed under the terms of the <a href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.</p>
</div>
</div>

<div id="rmd-source-code">---
title: "Day 1 - Module 03: Reference Genomes"
author: "UM Bioinformatics Core"
output:
        html_document:
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: true
            fig_caption: true
            markdown: GFM
            code_download: true
---
<style type="text/css">
body{ /* Normal  */
      font-size: 14pt;
  }
pre {
  font-size: 12pt
}
</style>

# Reference Genomes

In this module, we will learn:

* what a reference genome is and what it contains
* details about the FASTA and GTF formats
* to appreciate the differences in gene identifiers
* how to download a reference genome

# Differential Expression Workflow

Here we will set the stage for the next steps by discussing reference genomes, which are integral to genome alignments and gene/isoform quantification. Along the way we will touch on some things to be aware of.

![](images/wayfinder/wayfinder-02.png)
<br>
<br>
<br>
<br>

# Reference Genomes

A reference genome consists of the **reference sequence** and, optionally, any number of **genomic annotations** that describe attributes about that sequence. Examples of annotations include:

* Gene models consisting of the location and other information about genes.
* Variants consisting of the location of common or rare genetic variants, their alleles, and frequencies.
* Small RNAs consisting of the location and other information about various types of small RNAs.

Of particular relevance to us for this workshop are the reference sequence and gene models.

## Reference Sequence

Reference sequence is stored in [FASTA](https://en.wikipedia.org/wiki/FASTA_format) files. They are similar to FASTQ files in their storage of sequence information, but their format is a little different in a couple ways:

1. Records are separated by lines beginning with `>` instead of `@`.
2. Only the sequence is stored in a FASTA file, there is no notion of quality attached to the nucleotides.

```
>chrM
GATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCAT
TTGGTATTTTCGTCTGGGGGGTGTGCACGCGATAGCATTGCGAGACGCTG
GAGCCGGAGCACCCTATGTCGCAGTATCTGTCTTTGATTCCTGCCTCATT
CTATTATTTATCGCACCTACGTTCAATATTACAGGCGAACATACCTACTA
AAGTGTGTTAATTAATTAATGCTTGTAGGACATAATAATAACAATTGAAT
GTCTGCACAGCCGCTTTCCACACAGACATCATAACAAAAAATTTCCACCA
AACCCCCCCCTCCCCCCGCTTCTGGCCACAGCACTTAAACACATCTCTGC
CAAACCCCAAAAACAAAGAACCCTAACACCAGCCTAACCAGATTTCAAAT
TTTATCTTTAGGCGGTATGCACTTTTAACAGTCACCCCCCAACTAACACA
```

## Gene Models

Well-characterized organisms (e.g. human, mouse, zebrafish) have fairly mature gene models. These are stored in [GTF](https://uswest.ensembl.org/info/website/upload/gff.html) format, which gives location and other information about each gene feature. Below are two examples:


    chr1	unknown	exon	11874	12227	.	+	.	gene_id "DDX11L1"; gene_name "DDX11L1"; transcript_id "NR_046018"; tss_id "TSS16932";
    chr1	unknown	exon	12613	12721	.	+	.	gene_id "DDX11L1"; gene_name "DDX11L1"; transcript_id "NR_046018"; tss_id "TSS16932";
    chr1	unknown	exon	13221	14409	.	+	.	gene_id "DDX11L1"; gene_name "DDX11L1"; transcript_id "NR_046018"; tss_id "TSS16932";
    chr1	unknown	exon	14362	14829	.	-	.	gene_id "WASH7P"; gene_name "WASH7P"; transcript_id "NR_024540"; tss_id "TSS8568";


    1	havana	gene	11869	14409	.	+	.	gene_id "ENSG00000223972"; gene_version "5"; gene_name "DDX11L1"; gene_source "havana"; gene_biotype "transcribed_unprocessed_pseudogene";
    1	havana	transcript	11869	14409	.	+	.	gene_id "ENSG00000223972"; gene_version "5"; transcript_id "ENST00000456328"; transcript_version "2"; gene_name "DDX11L1"; gene_source "havana"; gene_biotype "transcribed_unprocessed_pseudogene"; transcript_name "DDX11L1-202"; transcript_source "havana"; transcript_biotype "lncRNA"; tag "basic"; transcript_support_level "1";
    1	havana	exon	11869	12227	.	+	.	gene_id "ENSG00000223972"; gene_version "5"; transcript_id "ENST00000456328"; transcript_version "2"; exon_number "1"; gene_name "DDX11L1"; gene_source "havana"; gene_biotype "transcribed_unprocessed_pseudogene"; transcript_name "DDX11L1-202"; transcript_source "havana"; transcript_biotype "lncRNA"; exon_id "ENSE00002234944"; exon_version "1"; tag "basic"; transcript_support_level "1";


The GTF format stores specific information in each column:

| Column | Description |
| :----: | ----------- |
| 1 | Chromosome |
| 2 | Source, e.g. ensembl, havana |
| 3 | Gene feature, e.g. exon, intron, mRNA, transcript |
| 4 | Start location, 1-based |
| 5 | End location, 1-based |
| 6 | Score |
| 7 | Strand |
| 8 | Frame, relating to codons |
| 9 | Attribute, a semicolon separated list of key/value pairs giving additional information about the feature. |

## Minutiae, Very Briefly

Bioinformatics is a relatively new, fast-changing, field and its data standards and formats are no different. Consequently there are some oddities and tedious items of note which we would like to only briefly touch on here.

### Genome Builds

On occassion new reference genomes are released, and the genome build number changes. You may be familiar with the UCSC manner of naming human genome builds: hg18, hg19, hg38. ENSEMBL, naturally, has their own way of referring to genome builds: GRCh36, GRCh37, and GRCh38. Notice with the most recent human reference, the numbering now aligns between UCSC and ENSEMBL.

Different organisms have their own versioning.

### Annotation Sources

[NCBI RefSeq](https://www.ncbi.nlm.nih.gov/refseq/rsg/), [ENSEMBL](https://www.ensembl.org/info/genome/genebuild/index.html), and [UCSC Known Genes](https://academic.oup.com/bioinformatics/article/22/9/1036/200093) are the three primary gene annotation databases (different organisms have their own databases). We will not go into exactly how the gene annotations are different, but we note that the are, and [others have examined the consequences of this](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-015-1308-8).

### Gene IDs

The two GTF examples above highlight different ways of referring to the same gene. In the first GTF we see:

* DDX11L1, the gene symbol, controlled by the [Human Gene Nomenclature Committee (HUGO)](https://www.genenames.org/).
* NR_046018, the RefSeq transcript ID

And in the second GTF we see:

* ENSG00000223972, the ENSEMBL gene ID
* DDX11L1, the gene symbol, thankfully the same
* ENST00000456328, the ENSEMBL transcript ID

Translating between different gene IDs is possible, as we will see in Day Two with `biomaRt`. But in terms of **best practice** it is generally a good idea to avoid using the gene symbol as the primary gene identifier because not everyone refers to the same gene by the same symbol.

# Getting a Reference Genome

The [Illumina iGenomes](https://support.illumina.com/sequencing/sequencing_software/igenome.html) resource is one of the easiest, and most comprehensive, ways to download a reference genome. iGenomes includes both the reference sequence and gene models.

Reference genomes can be **very large**, depending on the organism, and so we will not download one to the Amazon instance we are using for this workshop. We've included instructions for downloading these, in case you want to download these to the server where you intend to later do a similar RNA-seq analysis (e.g. on High-Performance Compute, GreatLakes).

<details>
<summary>How would I download references with iGenomes?</summary>

As noted, it's not recommended to download the iGenomes references to the AWS instance. However, if you wanted to know in general how you would do that, the process is described here.

First go to the [iGenomes](https://support.illumina.com/sequencing/sequencing_software/igenome.html) page, find the build you want from the source you want, right click the genome build you want to download, and select "Copy link location":

![iGenomes image for copying link location](images/genome_copy_link.png)

Then on the remote server you would go to the directory you'd like to download the genome to and type (that URL is what we copied):

```
$ wget http://igenomes.illumina.com.s3-website-us-east-1.amazonaws.com/Homo_sapiens/NCBI/GRCh38/Homo_sapiens_NCBI_GRCh38.tar.gz
```

After the download finishes (it may take a while as it is tens of GB large), you can unpack it with:

```
$ tar -xf Homo_sapiens_NCBI_GRCh38.tar.gz
```

</details>


## Which Reference is Right for Me?

The key is to be consistent in your research. Switching from ENSEMBL to UCSC will create many headaches because of the change in gene identifiers, and differences in the gene models themselves. Often people choose the one they're most comfortable with, which is often a function of historical accident. The key is not to overthink it.

Another important note is not to mix the sources. If you download reference sequence from UCSC, don't use an ENSEMBL GTF (and vice versa). One of the quirky differences between the two databases is that ENSEMBL refers to chromosome only by their number, i.e. `1`, whereas UCSC refers to chromsomes as `chr1`. This makes reference FASTAs from one source incompatible with gene builds from another.

---

These materials have been adapted and extended from materials created by the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module03_Reference_Genomes.md");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
