<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2021-09-16" />

<title>Day 2 - Module 08: Count normalizations and DESeq2 model fitting</title>

<script src="site_libs/header-attrs-2.10/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02_QC.html">Quality Control</a>
    </li>
    <li>
      <a href="Module02optional_Cutadapt_MultiQC.html">Optional: Cutadapt/MultiQC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module04_Alignment.html">Alignment/Quantification (STAR/RSEM)</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module05_MultiQC_and_Count_Matrix.html">Count matrix / Alignment QC</a>
    </li>
    <li>
      <a href="Module05optional_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">Introduction to DESeq2</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DESeq2 Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">Count Normalizations and DESeq2 model fitting</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample Visualizations for QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Comparisons and Gene Annotations</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Day 2 - Module 08: Count normalizations and DESeq2 model fitting</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2021-09-16</h4>

</div>


<!--- Allow the page to be wider --->
<style>
    body .main-container {
        max-width: 1200px;
    }
</style>
<blockquote>
<h1 id="objectives">Objectives:</h1>
<ul>
<li>Discuss count normalizations</li>
<li>Execute model fitting for differential expression comparisons</li>
</ul>
</blockquote>
<div id="differential-expression-workflow" class="section level1" number="1">
<h1 number="1"><span class="header-section-number">1</span> Differential Expression Workflow</h1>
<p>Here we will proceed with count normalizations and fit our DESeq2 model.</p>
<p><img src="images/wayfinder_07.png" style="width:75.0%" /></p>
<hr />
</div>
<div id="count-normalizations" class="section level1" number="2">
<h1 number="2"><span class="header-section-number">2</span> Count normalizations</h1>
<p>Since counts of mapped reads for each gene is proportional to the expression of RNA in addition to many “uninteresting” other factors, normalization is the process of scaling raw count values to account for the “uninteresting” factors and ensure expression levels are more comparable.</p>
<div id="normalization-goals" class="section level2" number="2.1">
<h2 number="2.1"><span class="header-section-number">2.1</span> Normalization goals</h2>
<p>Two common factors that need to be accounted for during normalization are <strong>sequencing depth</strong> and <strong>gene length</strong>. Common normalization approaches (such as FPKM, RPKM, CPM, TPM, etc.) account for one or both of these factors.</p>
<ul>
<li><strong>Sequencing depth</strong> normalization is neccessary to account for the proportion of reads per gene expected for more deeply sequenced samples (like in pink below) versus a less deeply sequenced sample (like in green below).</li>
</ul>
<div class="figure">
<img src="images/normalization_methods_depth.png" alt="" />
<p class="caption"><em>Note that each pink or green rectangle represents an aligned read, with reads spanning an intron connected by a dashed line. Figure from <a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">HBC training materials</a></em></p>
</div>
<ul>
<li><strong>Gene length</strong> normalization is necessary since genes of different lengths have different probablities of generating fragments that end up in the library. In the example below, both genes have similar levels of expression. However, the number of reads that map to the longer gene (Gene X) will be much great than the number of reads that map to the short gene (Gene Y).</li>
</ul>
<div class="figure">
<img src="images/normalization_methods_length.png" style="width:75.0%" alt="" />
<p class="caption"><em>Each pink rectangle represents an aligned read, with reads spanning an intron connected by a dashed line. Figure from <a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">HBC training materials</a></em></p>
</div>
<blockquote>
<p><strong>Note</strong>: The above figures are originally from a <a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">HBC tutorial</a> that also includes a detailed comparison of different normalization (CPM, TPM, FPKM) approaches and their best uses.</p>
</blockquote>
<p><strong>Check-in: Questions about normalizations?</strong></p>
</div>
<div id="deseq2-normalizations" class="section level2" number="2.2">
<h2 number="2.2"><span class="header-section-number">2.2</span> DESeq2 normalizations</h2>
<p>An additional consideration for normalization is <strong>RNA composition</strong>. A few highly differentially expressed genes, differences in the number of genes expressed between samples, or contamination are not accounted for by depth or gene length normalization methods. Accounting for RNA composition is particularly important for differential expression analyses, regardless of the tool used.</p>
<p>DESeq2 has an <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106">internal normalization process</a>. However for data exploration and visualizations, it is helpful to generate an object of independently normalized counts.</p>
<p>For downstream sample level quality control visualizations, we will use the <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations">rlog transformation</a> that accounts for sequencing depth for each sample and RNA composition differences in the overall experiment.</p>
<p>The rlog transformation produces log2 scale data that has also been normalized to overall library size as well as variance across genes at different mean expression levels. For larger numbers of samples, there is an alternative transformation method, <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations">vst</a> that can be used instead for count normalizations.</p>
<p>The command to generate the normalized count object has a few parts, including <code>dds</code> as an input and providing a value to the option <code>blind</code>.</p>
<pre class="r"><code>rld &lt;- rlog(dds, blind = TRUE)</code></pre>
<pre><code>## -- note: fitType=&#39;parametric&#39;, but the dispersion trend was not well captured by the
##    function: y = a/x + b, and a local regression fit was automatically substituted.
##    specify fitType=&#39;local&#39; or &#39;mean&#39; to avoid this message next time.</code></pre>
<p>*Note: We see a message here that our data did not fit the default ‘parametric’ dispersion model so a local regression was substituted. If we had more time, we might look at a dispersion plot with the <code>plotDispEsts(dds)</code> function, but as <a href="https://support.bioconductor.org/p/107937/">this bioconductor thread</a> discusses, other visualizations of our data might be more helpful and/or easier to interpret.</p>
<p>Next, we’ll look at the results of the tranformation.</p>
<pre class="r"><code>head(assay(rld), 2)</code></pre>
<pre><code>##                    Sample_116498 Sample_116499 Sample_116500 Sample_116501 Sample_116502
## ENSMUSG00000000001     12.639049     12.609976     12.791192     12.619063     12.721280
## ENSMUSG00000000028      7.847671      8.152787      7.949574      8.203993      8.464987
##                    Sample_116503 Sample_116504 Sample_116505 Sample_116506 Sample_116507
## ENSMUSG00000000001     12.668801      12.53167      12.50798      12.47565      12.72717
## ENSMUSG00000000028      8.506084      10.82552      10.89710      11.00659      11.15996
##                    Sample_116508 Sample_116509
## ENSMUSG00000000001      12.51363      12.46995
## ENSMUSG00000000028      11.06722      11.09489</code></pre>
<p>Looking at the rld values, we can see that they are now in log scale. Since we set <code>blind=TRUE</code>, the transformation is blind to the sample information we specified in the design formula. The normalized counts are helpful for visualization methods during expresion-level quality assessment but <strong>aren’t used in the model fitting</strong>.</p>
</div>
</div>
<div id="deseq2-model-fitting" class="section level1" number="3">
<h1 number="3"><span class="header-section-number">3</span> DESeq2 Model Fitting</h1>
<p>Next, we’ll fit our standard model using the <code>DESeq</code> function and take a look at the objects we generate.</p>
<pre class="r"><code>dds &lt;- DESeq(dds)</code></pre>
<p>This command applies the model to our data, using the sample information supplied when generating the <code>dds</code> object so it takes some time to run.</p>
<pre class="r"><code>resultsNames(dds)</code></pre>
<pre><code>## [1] &quot;Intercept&quot;                    &quot;Gtype.Tx_ko.Tx_vs_wt.Tx&quot;     
## [3] &quot;Gtype.Tx_ko.control_vs_wt.Tx&quot; &quot;Gtype.Tx_wt.control_vs_wt.Tx&quot;</code></pre>
<p>The results include three pairwise comparisons to the specified control as default but other information is now stored in the <code>dds</code> object so we can generate additional pairwise comparisons.</p>
<pre class="r"><code>head(dds)</code></pre>
<p><strong>Checkpoint</strong>: <em>If you see the same results when you execute <code>resultsNames(dds)</code>, please indicate with the green ‘yes’ button. Otherwise, please use the red ‘x’ button to get help before the break</em></p>
<div id="optional-exercise" class="section level4" number="3.0.0.1">
<h4 number="3.0.0.1"><span class="header-section-number">3.0.0.1</span> [Optional Exercise]</h4>
<p>If you executed the commands in the option exercise from Module 07, you can fit a seperate DESeq2 model for the patient sample example, which included a covariate in our model.</p>
<details>
<summary>
<em>Click for fitting a model that includes a covariate</em>
</summary>
<p><code>r     dds_patient &lt;- DESeq(dds_patient)</code></p>
<code>r     resultsNames(dds_patient)</code> If you run through the optional exercises, you can explore the impact of adding a covariate by substituting <code>dds_patient</code> for <code>dds</code> and re-running those commands since both DESeq2 objects have their data organized in the same way.
</details>
</div>
</div>
<div id="summary" class="section level1" number="4">
<h1 number="4"><span class="header-section-number">4</span> Summary</h1>
<p>In this section, we:</p>
<ul>
<li>Learned about count normalizations and uses<br />
</li>
<li>Generated a normalized count table<br />
</li>
<li>Fit two DESeq2 models for our data<br />
</li>
<li>Saw the impact of including a covariate in our model</li>
</ul>
<p><strong>Before moving on with our group comparisons, we’ll take a short (10 minute) break.</strong></p>
<hr />
</div>
<div id="sources" class="section level1" number="5">
<h1 number="5"><span class="header-section-number">5</span> Sources</h1>
<div id="training-resources-used-to-develop-materials" class="section level2" number="5.1">
<h2 number="5.1"><span class="header-section-number">5.1</span> Training resources used to develop materials</h2>
<ul>
<li>HBC DGE setup: <a href="https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html" class="uri">https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html</a></li>
<li>HBC Count Normalization: <a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html" class="uri">https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html</a></li>
<li>DESeq2 standard vignette: <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html" class="uri">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html</a></li>
<li>DESeq2 beginners vignette: <a href="https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf" class="uri">https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf</a></li>
<li>Bioconductor RNA-seq Workflows: <a href="https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html" class="uri">https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html</a></li>
</ul>
<p>These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the <a href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.</p>
</div>
</div>

<div id="rmd-source-code">---
title: "Day 2 - Module 08: Count normalizations and DESeq2 model fitting"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: true
            fig_caption: true
            markdown: GFM
            code_download: true
---

<!--- Allow the page to be wider --->
<style>
    body .main-container {
        max-width: 1200px;
    }
</style>

> # Objectives:    
> * Discuss count normalizations
> * Execute model fitting for differential expression comparisons

```{r LoadRunningData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
load("rdata/RunningData.RData")
```


# Differential Expression Workflow

Here we will proceed with count normalizations and fit our DESeq2 model.

![](./images/wayfinder_07.png){width=75%}

---

# Count normalizations

Since counts of mapped reads for each gene is proportional to the expression of RNA in addition to many “uninteresting” other factors, normalization is the process of scaling raw count values to account for the “uninteresting” factors and ensure expression levels are more comparable.

## Normalization goals

Two common factors that need to be accounted for during normalization are **sequencing depth** and **gene length**. Common normalization approaches (such as FPKM, RPKM, CPM, TPM, etc.) account for one or both of these factors.    

* **Sequencing depth** normalization is neccessary to account for the proportion of reads per gene expected for more deeply sequenced samples (like in pink below) versus a less deeply sequenced sample (like in green below).    

![*Note that each pink or green rectangle represents an aligned read, with reads spanning an intron connected by a dashed line. Figure from [HBC  training materials](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html)*](./images/normalization_methods_depth.png)    

    
* **Gene length** normalization is necessary since genes of different lengths have different probablities of generating fragments that end up in the library. In the example below, both genes have similar levels of expression. However, the number of reads that map to the longer gene (Gene X) will be much great than the number of reads that map to the short gene (Gene Y).    

![*Each pink rectangle represents an aligned read, with reads spanning an intron connected by a dashed line. Figure from [HBC  training materials](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html)*](./images/normalization_methods_length.png){width=75%}      
    

> **Note**: The above figures are originally from a [HBC  tutorial](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html) that also includes a detailed comparison of different normalization (CPM, TPM, FPKM) approaches and their best uses.

**Check-in: Questions about normalizations?**

## DESeq2 normalizations

An additional consideration for normalization is **RNA composition**. A few highly differentially expressed genes, differences in the number of genes expressed between samples, or contamination are not accounted for by depth or gene length normalization methods. Accounting for RNA composition is particularly important for differential expression analyses, regardless of the tool used.

DESeq2 has an [internal normalization process](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106). However for data exploration and visualizations, it is helpful to generate an object of independently normalized counts.

For downstream sample level quality control visualizations, we will use the [rlog transformation](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations) that accounts for sequencing depth for each sample and RNA composition differences in the overall experiment.

The rlog transformation produces log2 scale data that has also been normalized to overall library size as well as variance across genes at different mean expression levels. For larger numbers of samples, there is an alternative transformation method, [vst](http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations) that can be used instead for count normalizations. 


The command to generate the normalized count object has a few parts, including `dds` as an input and providing a value to the option `blind`.
```{r CountNorm}
rld <- rlog(dds, blind = TRUE)
```

*Note: We see a message here that our data did not fit the default 'parametric' dispersion model so a local regression was substituted. If we had more time, we might look at a dispersion plot with the `plotDispEsts(dds)` function, but as [this bioconductor thread](https://support.bioconductor.org/p/107937/) discusses, other visualizations of our data might be more helpful and/or easier to interpret. 

Next, we'll look at the results of the tranformation.
```{r CountNormCheck}
head(assay(rld), 2)
```

Looking at the rld values, we can see that they are now in log scale. Since we set `blind=TRUE`, the transformation is blind to the sample information we specified in the design formula. The normalized counts are helpful for visualization methods during expresion-level quality assessment but **aren't used in the model fitting**. 

# DESeq2 Model Fitting

Next, we'll fit our standard model using the `DESeq` function and take a look at the objects we generate. 
```{r FitModelStandard, message=FALSE}
dds <- DESeq(dds)
```

This command applies the model to our data, using the sample information supplied when generating the `dds` object so it takes some time to run. 

```{r FitModelStandardCheck1, eval=TRUE}
resultsNames(dds)
```

The results include three pairwise comparisons to the specified control as default but other information is now stored in the `dds` object so we can generate additional pairwise comparisons.

```{r FitModelStandardCheck2, eval=FALSE}
head(dds)
```

**Checkpoint**: *If you see the same results when you execute `resultsNames(dds)`, please indicate with the green 'yes' button. Otherwise, please use the red 'x' button to get help before the break*



#### [Optional Exercise]

If you executed the commands in the option exercise from Module 07, you can fit a seperate DESeq2 model for the patient sample example, which included a covariate in our model. 

<details>
    <summary>*Click for fitting a model that includes a covariate*</summary>
    ```{r FitModelCovariate, eval= FALSE, message=FALSE}
    dds_patient <- DESeq(dds_patient)
    ```
    ```{r FitModelCovariateCheck1,eval=FALSE}
    resultsNames(dds_patient)
    ```
    If you run through the optional exercises, you can explore the impact of adding a covariate by substituting `dds_patient` for `dds` and re-running those commands since both DESeq2 objects have their data organized in the same way.
</details>




# Summary 

In this section, we:   

* Learned about count normalizations and uses    
* Generated a normalized count table    
* Fit two DESeq2 models for our data    
* Saw the impact of including a covariate in our model 

**Before moving on with our group comparisons, we'll take a short (10 minute) break.**


---

# Sources
## Training resources used to develop materials
* HBC DGE setup: https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html
* HBC Count Normalization: https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html
* DESeq2 standard vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
* DESeq2 beginners vignette: https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf
* Bioconductor RNA-seq Workflows: https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html



```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
#Hidden code block to write out data for knitting
save.image(file = "rdata/RunningData.RData")

```


These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module08_DESeq2DE.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
