<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />


<title>Day 1 - Module 02: QC</title>

<script src="site_libs/header-attrs-2.10/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02_QC.html">Quality Control</a>
    </li>
    <li>
      <a href="Module02optional_Cutadapt_MultiQC.html">Optional: Cutadapt/MultiQC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module04_Alignment.html">Alignment/Quantification (STAR/RSEM)</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module05_MultiQC_and_Count_Matrix.html">Count matrix / Alignment QC</a>
    </li>
    <li>
      <a href="Module05optional_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">Introduction to DESeq2</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DESeq2 Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">Count Normalizations and DESeq2 model fitting</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample Visualizations for QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Comparisons and Gene Annotations</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Day 1 - Module 02: QC</h1>
<h4 class="author">UM Bioinformatics Core</h4>

</div>


<div id="qc" class="section level1" number="1">
<h1 number="1"><span class="header-section-number">1</span> QC</h1>
<p>In this Quality Control (QC) module, we will learn:</p>
<ul>
<li>what a FASTQ file is</li>
<li>how to use FastQC to assess the quality of sequencing reads</li>
<li>how to interpret FastQC reports</li>
</ul>
</div>
<div id="differential-expression-workflow" class="section level1" number="2">
<h1 number="2"><span class="header-section-number">2</span> Differential Expression Workflow</h1>
<p>An overview of the steps of entire differential expression experiment and analysis workflow is shown below. In this lesson, we will go over the highlighted part of the workflow.</p>
<p><img src="images/wayfinder/wayfinder-01.png" /> <br> <br> <br> <br></p>
</div>
<div id="rna-seq-data" class="section level1" number="3">
<h1 number="3"><span class="header-section-number">3</span> RNA-seq Data</h1>
<p>Analysis of RNA-seq requires the use of many types of data. The three most common that we will encounter are:</p>
<ul>
<li>Sequence data</li>
<li>Genome feature data</li>
<li>Alignment data</li>
</ul>
<p>We will introduce each data type and their formats in more detail as they enter the workflow, but for now some general notes about the formats we will encounter:</p>
<ul>
<li>File formats represent the data in a structured manner.</li>
<li>Some file formats are not human-readable; they are binary, and we will see what that means when we discuss alignments.</li>
<li>Many file formats are human readable, but very large. <strong>These files should not be viewed in Word or Excel.</strong></li>
</ul>
<div id="fastq-files" class="section level2" number="3.1">
<h2 number="3.1"><span class="header-section-number">3.1</span> FASTQ Files</h2>
<p>Raw sequenced reads are stored in <a href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ</a> format. This is a plain text, human readable file format that stores information about each read. For a single read there are four lines:</p>
<pre><code>@A00437:266:HVFMYDSXX:1:1101:1181:1000 1:N:0:AGCCTATC+TGCGTAAC
TTTATTGTTGATGGTTATTTTTTGTTTATGGTTATTTTGTGTTTATGGTTATTTTTTGTTTATGGTTATTTTTTA
+
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF</code></pre>
<table>
<colgroup>
<col width="26%" />
<col width="73%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Line</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">1</td>
<td>Begins with <code>@</code> and has information about the read. From AGC will contain flowcell and barcode information.</td>
</tr>
<tr class="even">
<td align="center">2</td>
<td>The RNA sequence.</td>
</tr>
<tr class="odd">
<td align="center">3</td>
<td>Always begins with <code>+</code> and may contain the same information as line 1.</td>
</tr>
<tr class="even">
<td align="center">4</td>
<td>A string of characters representing the quality scores. Always the same number of characters as line 2.</td>
</tr>
</tbody>
</table>
<div id="base-quality-scores" class="section level3" number="3.1.1">
<h3 number="3.1.1"><span class="header-section-number">3.1.1</span> Base Quality Scores</h3>
<p>The fourth line encodes the quality of each nucleotide in the read. The most common encoding is referred to as PHRED-33, and the following maps characters to qualities:</p>
<pre><code>Quality encoding: !&quot;#$%&amp;&#39;()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHI
                  |         |         |         |         |
   Quality score: 0........10........20........30........40</code></pre>
<p>The above example is a rather boring read with all bases having <code>F</code>, meaning a score of 37, which is good. Each quality score represents the probability that the corresponding nucleotide call is incorrect. The quality score is logarithmic and is calculated as:</p>
<pre><code>Q = -10 x log10(P), where P is the probability that the base call is erroneous</code></pre>
<p>Another way to interpret the scores is in terms of the following table:</p>
<table>
<thead>
<tr class="header">
<th>PHRED Quality Score</th>
<th>Probability of incorrect base call</th>
<th>Base call accuracy</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>10</td>
<td>1 in 10</td>
<td>90%</td>
</tr>
<tr class="even">
<td>20</td>
<td>1 in 100</td>
<td>99%</td>
</tr>
<tr class="odd">
<td>30</td>
<td>1 in 1000</td>
<td>99.9%</td>
</tr>
<tr class="even">
<td>40</td>
<td>1 in 10000</td>
<td>99.99%</td>
</tr>
</tbody>
</table>
<p>It would clearly be painstaking to go through all the base quality scores of all the millions of reads obtained from a sequencing run, so we typically don’t evaluate the FASTQ files by sight. Instead, we use tools that take in FASTQ files and summarize the base quality scores (and other characteristics) to determine the overall quality of the reads from the sequencer.</p>
</div>
</div>
</div>
<div id="quality-assessment-with-fastqc" class="section level1" number="4">
<h1 number="4"><span class="header-section-number">4</span> Quality Assessment with FastQC</h1>
<div class="figure">
<img src="images/fastqc_logo.png" alt="" />
<p class="caption">FastQC</p>
</div>
<p>The first step in many high-throughput sequencing analyses is to examine the quality of the raw reads with <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>. It analyzes various aspects of each FASTQ file and provides a quick impression of the overall quality metrics which are useful to determine before proceeding with downstream analysis. The following modules are present in an easy to read HTML report.</p>
<ul>
<li>Read quality score per base (by interpreting the PHRED33 scores)</li>
<li>Per tile sequencing quality (related to sequencing clusters on the flowcell)</li>
<li>Read lengths</li>
<li>GC content</li>
<li>Adapter content</li>
<li>Overrepresented sequences</li>
</ul>
<p>Toward the end of this module, we’ll run FastQC. For now, we’ll discuss the output files it produces and view some examples.</p>
<div id="fastqc-output-files" class="section level2" number="4.1">
<h2 number="4.1"><span class="header-section-number">4.1</span> FastQC output files</h2>
<p>FastQC outputs an HTML report and a zipped archive of intermediate files. These will appear like this:</p>
<pre><code>sample_01_R1.trimmed_fastqc.html
sample_01_R1.trimmed_fastqc.zip</code></pre>
<p>For most situations, opening the HTML report and inspecting the results will provide the information necessary to make decisions about your data.</p>
<p>First we’ll take a look at some of the figures found in a typical FastQC report, then we’ll view an example of an HTML report and try to make interpretations from this. Finally, we’ll run FastQC to produce some of these files from our example dataset.</p>
</div>
<div id="interpreting-fastqc" class="section level2" number="4.2">
<h2 number="4.2"><span class="header-section-number">4.2</span> Interpreting FastQC</h2>
<p>FastQC reports are organized around the the different modules characterizing different aspects of the reads. The FastQC <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/">help page</a> has detailed descriptions for each module.</p>
<p>The report summary is a PASS, WARN, FAIL overview of the modules. It’s important to not take them too seriously in all cases as there are a number of assumptions that go into the thresholds that may not apply in your particular experiment.</p>
<p>An important note: For computational reasons, FastQC only evaluates the first 100,000 reads in a FASTQ file. The assumption is that the first 100,000 reads sufficiently represent the full library.</p>
<div id="per-base-sequence-quality" class="section level3" number="4.2.1">
<h3 number="4.2.1"><span class="header-section-number">4.2.1</span> Per base sequence quality</h3>
<p>A per base boxplot of the quality scores with helpful coloring for good, fair, and poor qualities. If a sample has quality scores that degrade at the 3’ end of a read (as in the second example below), the sample doesn’t necessarily have to be dropped. Some aligners, e.g. STAR, will automatically trim poor quality read ends.</p>
<center>
<p><img src="images/fastqc_per_base.png" width="800"/></p>
<p>An example of a good quality run.</p>
</center>
<center>
<p><img src="images/fastqc_poor_per_base.png" width="800" /></p>
<p>An example of a poor quality run.</p>
</center>
</div>
<div id="sequence-duplication" class="section level3" number="4.2.2">
<h3 number="4.2.2"><span class="header-section-number">4.2.2</span> Sequence Duplication</h3>
<p>High sequence duplication, indicated by a skew to the right of the blue curve in the plots below, can be caused by low complexity libraries, too many cycles of PCR amplification, or too little starting material.</p>
<p>On the other hand, if the library was prepared with an enrichment step (e.g. ChIP-seq), then sequence duplication might not be unexpected.</p>
<center>
<p><img src="images/fastqc_duplication.png" width="800" /></p>
<p>An example of a sample with low sequence duplication levels.</p>
</center>
<center>
<p><img src="images/fastqc_poor_duplication.png" width="800" /></p>
<p>An example of a sample with high sequence duplication levels.</p>
</center>
</div>
<div id="overrepresented-sequences" class="section level3" number="4.2.3">
<h3 number="4.2.3"><span class="header-section-number">4.2.3</span> Overrepresented sequences</h3>
<p>Overrepresented sequences are those making up more than 0.1% of the total (of the 100,000 reads sampled). Finding an overrepresented sequence is an indication of something biologically meaningful (e.g. an enrichment step in the library prep process), or it is an issue stemming from the library preparation process (e.g. adapters, a failure to fully remove polyA tails, or contamination).</p>
<center>
<p><img src="images/fastqc_overrepresented.png" width="800" /></p>
<p>We see that there is a high number of reads with polyA tails. This could be a result of an inefficient removal of polyA tails.</p>
</center>
</div>
<div id="adapter-content" class="section level3" number="4.2.4">
<h3 number="4.2.4"><span class="header-section-number">4.2.4</span> Adapter Content</h3>
<p>Adapter sequence typically is not present in standard RNA-seq libraries. However, if the fragment sizes resulting from library prep are smaller than the read length on the sequencer, then it is possible to get “read-through” of the adapter.</p>
<p>This can be dealt with by adding a step to the workflow that trims the reads of adapter content with a tool such as <a href="https://cutadapt.readthedocs.io/en/stable/">cutadapt</a>. We will see this in a later module.</p>
</div>
</div>
<div id="examining-a-fastqc-report" class="section level2" number="4.3">
<h2 number="4.3"><span class="header-section-number">4.3</span> Examining a FastQC report</h2>
<p>From the zipped folder you downloaded earlier, find the FastQC report <code>sample_01_R1.trimmed_fastqc.html</code>.</p>
<p>We’ll view this report and try to interpret some of the figures.</p>
</div>
<div id="acting-on-qc---good-or-bad-results" class="section level2" number="4.4">
<h2 number="4.4"><span class="header-section-number">4.4</span> Acting on QC - Good or Bad Results?</h2>
<p>Good QC on raw reads means that RNA quality, library preparation, and sequencing seem fine to proceed with the next steps of analysis. It <strong>does not</strong> mean that you will find differentially expressed genes downstream.</p>
<p>The source of bad QC measures are not always easy to determine, and doesn’t necessarily mean the data cannot be used downstream. However, bad QC measures may indicate that hazards lie ahead, and that certain steps upstream of sequencing (sample handling, RNA extraction, library preparation) might need to be altered.</p>
</div>
<div id="fastqc-exercise" class="section level2" number="4.5">
<h2 number="4.5"><span class="header-section-number">4.5</span> FastQC Exercise</h2>
<p>Let’s try running FastQC on our FASTQ input files.</p>
<ol style="list-style-type: decimal">
<li>Navigate to <code>analysis</code> directory</li>
<li>View FastQC help file</li>
<li>Construct the call to FastQC</li>
<li>Watch FastQC process the files / gather quality metrics</li>
<li>View the output of FastQC (the filenames)</li>
</ol>
<pre><code># Navigate to analysis directory
cd ~/analysis
# View the FastQC help file
fastqc -h
# Contstruct the call to FastQC
fastqc -o out_fastqc/ ../data/reads/*.fastq.gz
# Watch the output of FastQC
# View the output of FastQC (the filenames)</code></pre>
<hr />
<p>These materials have been adapted and extended from materials created by the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. These are open access materials distributed under the terms of the <a href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.</p>
</div>
</div>

<div id="rmd-source-code">---
title: "Day 1 - Module 02: QC"
author: "UM Bioinformatics Core"
output:
        html_document:
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: true
            fig_caption: true
            markdown: GFM
            code_download: true
---

# QC

In this Quality Control (QC) module, we will learn:

* what a FASTQ file is
* how to use FastQC to assess the quality of sequencing reads
* how to interpret FastQC reports

# Differential Expression Workflow

An overview of the steps of entire differential expression experiment and analysis workflow is shown below. In this lesson, we will go over the highlighted part of the workflow.

![](images/wayfinder/wayfinder-01.png)
<br>
<br>
<br>
<br>

# RNA-seq Data

Analysis of RNA-seq requires the use of many types of data. The three most common that we will encounter are:

- Sequence data
- Genome feature data
- Alignment data

We will introduce each data type and their formats in more detail as they enter the workflow, but for now some general notes about the formats we will encounter:

- File formats represent the data in a structured manner.
- Some file formats are not human-readable; they are binary, and we will see what that means when we discuss alignments.
- Many file formats are human readable, but very large. **These files should not be viewed in Word or Excel.**

## FASTQ Files

Raw sequenced reads are stored in [FASTQ](https://en.wikipedia.org/wiki/FASTQ_format) format. This is a plain text, human readable file format that stores information about each read. For a single read there are four lines:


    @A00437:266:HVFMYDSXX:1:1101:1181:1000 1:N:0:AGCCTATC+TGCGTAAC
    TTTATTGTTGATGGTTATTTTTTGTTTATGGTTATTTTGTGTTTATGGTTATTTTTTGTTTATGGTTATTTTTTA
    +
    FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF


| Line | Description |
| :--: | ----------- |
| 1 | Begins with `@` and has information about the read. From AGC will contain flowcell and barcode information. |
| 2 | The RNA sequence. |
| 3 | Always begins with `+` and may contain the same information as line 1. |
| 4 | A string of characters representing the quality scores. Always the same number of characters as line 2. |

### Base Quality Scores

The fourth line encodes the quality of each nucleotide in the read. The most common encoding is referred to as PHRED-33, and the following maps characters to qualities:


    Quality encoding: !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHI
                      |         |         |         |         |
       Quality score: 0........10........20........30........40


The above example is a rather boring read with all bases having `F`, meaning a score of 37, which is good. Each quality score represents the probability that the corresponding nucleotide call is incorrect. The quality score is logarithmic and is calculated as:


    Q = -10 x log10(P), where P is the probability that the base call is erroneous


Another way to interpret the scores is in terms of the following table:

| PHRED Quality Score | Probability of incorrect base call | Base call accuracy |
| ------------------- | ---------------------------------- | ------------------ |
| 10 | 1 in 10 | 90% |
| 20 | 1 in 100 | 99% |
| 30 | 1 in 1000 | 99.9% |
| 40 | 1 in 10000 | 99.99% |

It would clearly be painstaking to go through all the base quality scores of all the millions of reads obtained from a sequencing run, so we typically don't evaluate the FASTQ files by sight. Instead, we use tools that take in FASTQ files and summarize the base quality scores (and other characteristics) to determine the overall quality of the reads from the sequencer.

# Quality Assessment with FastQC

![FastQC](images/fastqc_logo.png)

The first step in many high-throughput sequencing analyses is to examine the quality of the raw reads with [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/). It analyzes various aspects of each FASTQ file and provides a quick impression of the overall quality metrics which are useful to determine before proceeding with downstream analysis. The following modules are present in an easy to read HTML report.

- Read quality score per base (by interpreting the PHRED33 scores)
- Per tile sequencing quality (related to sequencing clusters on the flowcell)
- Read lengths
- GC content
- Adapter content
- Overrepresented sequences

Toward the end of this module, we'll run FastQC. For now, we'll discuss the output files it produces and view some examples.

## FastQC output files

FastQC outputs an HTML report and a zipped archive of intermediate files. These will appear like this:

    sample_01_R1.trimmed_fastqc.html
    sample_01_R1.trimmed_fastqc.zip

For most situations, opening the HTML report and inspecting the results will provide the information necessary to make decisions about your data.

First we'll take a look at some of the figures found in a typical FastQC report, then we'll view an example of an HTML report and try to make interpretations from this. Finally, we'll run FastQC to produce some of these files from our example dataset.


## Interpreting FastQC

FastQC reports are organized around the the different modules characterizing different aspects of the reads. The FastQC [help page](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/) has detailed descriptions for each module.

The report summary is a PASS, WARN, FAIL overview of the modules. It's important to not take them too seriously in all cases as there are a number of assumptions that go into the thresholds that may not apply in your particular experiment.

An important note: For computational reasons, FastQC only evaluates the first 100,000 reads in a FASTQ file. The assumption is that the first 100,000 reads sufficiently represent the full library.

### Per base sequence quality

A per base boxplot of the quality scores with helpful coloring for good, fair, and poor qualities. If a sample has quality scores that degrade at the 3' end of a read (as in the second example below), the sample doesn't necessarily have to be dropped. Some aligners, e.g. STAR, will automatically trim poor quality read ends.

<center>

<img src="images/fastqc_per_base.png" width="800"/>


An example of a good quality run.

</center>

<center>

<img src="images/fastqc_poor_per_base.png" width="800" />

An example of a poor quality run.

</center>

### Sequence Duplication

High sequence duplication, indicated by a skew to the right of the blue curve in the plots below, can be caused by low complexity libraries, too many cycles of PCR amplification, or too little starting material.

On the other hand, if the library was prepared with an enrichment step (e.g. ChIP-seq), then sequence duplication might not be unexpected.

<center>

<img src="images/fastqc_duplication.png" width="800" />

An example of a sample with low sequence duplication levels.

</center>

<center>

<img src="images/fastqc_poor_duplication.png" width="800" />

An example of a sample with high sequence duplication levels.

</center>

### Overrepresented sequences

Overrepresented sequences are those making up more than 0.1% of the total (of the 100,000 reads sampled). Finding an overrepresented sequence is an indication of something biologically meaningful (e.g. an enrichment step in the library prep process), or it is an issue stemming from the library preparation process (e.g. adapters, a failure to fully remove polyA tails, or contamination).

<center>

<img src="images/fastqc_overrepresented.png" width="800" />

We see that there is a high number of reads with polyA tails. This could be a result of an inefficient removal of polyA tails.

</center>

### Adapter Content

Adapter sequence typically is not present in standard RNA-seq libraries. However, if the fragment sizes resulting from library prep are smaller than the read length on the sequencer, then it is possible to get "read-through" of the adapter.

This can be dealt with by adding a step to the workflow that trims the reads of adapter content with a tool such as [cutadapt](https://cutadapt.readthedocs.io/en/stable/). We will see this in a later module.

## Examining a FastQC report

From the zipped folder you downloaded earlier, find the FastQC report `sample_01_R1.trimmed_fastqc.html`.

We'll view this report and try to interpret some of the figures.

## Acting on QC - Good or Bad Results?

Good QC on raw reads means that RNA quality, library preparation, and sequencing seem fine to proceed with the next steps of analysis. It **does not** mean that you will find differentially expressed genes downstream.

The source of bad QC measures are not always easy to determine, and doesn't necessarily mean the data cannot be used downstream. However, bad QC measures may indicate that hazards lie ahead, and that certain steps upstream of sequencing (sample handling, RNA extraction, library preparation) might need to be altered.

## FastQC Exercise

Let's try running FastQC on our FASTQ input files.

1. Navigate to `analysis` directory
2. View FastQC help file
3. Construct the call to FastQC
4. Watch FastQC process the files / gather quality metrics
5. View the output of FastQC (the filenames)

```
# Navigate to analysis directory
cd ~/analysis
# View the FastQC help file
fastqc -h
# Contstruct the call to FastQC
fastqc -o out_fastqc/ ../data/reads/*.fastq.gz
# Watch the output of FastQC
# View the output of FastQC (the filenames)
```

---

These materials have been adapted and extended from materials created by the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module02_QC.md");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
